Libraries/LibTLS/Record.cpp

```diff
@@ -0,0 +1,316 @@
+        return (i8)Error::NeedMoreData;
+    }
+
+    auto type = (MessageType)buffer[0];
+    size_t buffer_position { 1 };
+
+    // FIXME: Read the version and verify it
+#ifdef TLS_DEBUG
+    auto version = (Version) * (const u16*)buffer.offset_pointer(buffer_position);
+    dbg() << "type: " << (u8)type << " version: " << (u16)version;
+#endif
+    buffer_position += 2;
+
+    auto length = convert_between_host_and_network(*(const u16*)buffer.offset_pointer(buffer_position));
+#ifdef TLS_DEBUG
+    dbg() << "record length: " << length << " at offset: " << buffer_position;
+#endif
+    buffer_position += 2;
+
+    if (buffer_position + length > buffer.size()) {
+        dbg() << "record length more than what we have: " << buffer.size();
+        return (i8)Error::NeedMoreData;
+    }
+
+#ifdef TLS_DEBUG
+    dbg() << "message type: " << (u8)type << ", length: " << length;
+#endif
+    ByteBuffer plain = buffer.slice_view(buffer_position, buffer.size() - buffer_position);
+
+    if (m_context.cipher_spec_set && type != MessageType::ChangeCipher) {
+#ifdef TLS_DEBUG
+        dbg() << "Encrypted: ";
+        print_buffer(buffer.slice_view(header_size, length));
+#endif
+
+        ASSERT(m_aes_remote);
+        auto iv_size = iv_length();
+
+        auto decrypted = m_aes_remote->create_aligned_buffer(length - iv_size);
+        auto iv = buffer.slice_view(header_size, iv_size);
+
+        m_aes_remote->decrypt(buffer.slice_view(header_size + iv_size, length - iv_size), decrypted, iv);
+
+        length = decrypted.size();
+
+#ifdef TLS_DEBUG
+        dbg() << "Decrypted: ";
+        print_buffer(decrypted);
+#endif
+
+        auto mac_size = mac_length();
+        if (length < mac_size) {
+            dbg() << "broken packet";
+            auto packet = build_alert(true, (u8)AlertDescription::DecryptError);
+            write_packet(packet);
+            return (i8)Error::BrokenPacket;
+        }
+
+        const u8* message_hmac = decrypted.offset_pointer(length - mac_size);
+        u8 temp_buf[5];
+        memcpy(temp_buf, buffer.offset_pointer(0), 3);
+        *(u16*)(temp_buf + 3) = convert_between_host_and_network(length);
+        auto hmac = hmac_message(ByteBuffer::wrap(temp_buf, 5), decrypted, mac_size);
+        auto message_mac = ByteBuffer::wrap(message_hmac, mac_size);
+        if (hmac != message_mac) {
+            dbg() << "integrity check failed (mac length " << length << ")";
+            dbg() << "mac received:";
+            print_buffer(message_mac);
+            dbg() << "mac computed:";
+            print_buffer(hmac);
+            auto packet = build_alert(true, (u8)AlertDescription::BadRecordMAC);
+            write_packet(packet);
+
+            return (i8)Error::IntegrityCheckFailed;
+        }
+        plain = decrypted.slice(0, length - mac_size);
+    }
+    m_context.remote_sequence_number++;
+
+    switch (type) {
+    case MessageType::ApplicationData:
+        if (m_context.connection_status != 0xff) {
```

Comment:
Styliing, feel free to ignore : Here and below in ChangeCipher, we're checking `connection_status` against magic values. Maybe add constants and/or `#DEFINE`s about it ?

Reply:
Absolutely.
I just spent non-zero time thinking what each value meant, putting them in an enum is very much the way to go